feat: add `blas/base/zdotu`

gururaj1512 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves a part of  #2039.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds javascript implementation for `blas/base/zdotu`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Resolves a part of  #2039.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add `blas/base/zdotu`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
bench: change variable naming

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
docs: improve jsdoc and change variable naming

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-blas-base-zdotu

ShabiShett07 committed:
chore: update repl

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: add examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
test: change variable naming

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update markdown file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,216 @@
+
+// Create offset views...
+var x1 = new Complex128Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
+var y1 = new Complex128Array( y0.buffer, y0.BYTES_PER_ELEMENT*2 ); // start at 3th element
```
```suggestion
var y1 = new Complex128Array( y0.buffer, y0.BYTES_PER_ELEMENT*2 ); // start at 3rd element
```

kgryte committed:
docs: fix comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented:
/stdlib merge

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,216 @@
+-   **y**: input [`Complex128Array`][@stdlib/array/complex128].
+-   **strideY**: stride kength for `y`.
+
+The `N` and strides parameters determine which elements in the strided arrays are accessed at runtime. For example, to calculate the dot product of every other value in `x` and the first `N` elements of `y` in reverse order,
```
```suggestion
The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to calculate the dot product of every other element in `x` and the first `N` elements of `y` in reverse order,
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,216 @@
+-   **offsetX**: starting index for `x`.
+-   **offsetY**: starting index for `y`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the dot product of every other value in `x` starting from the second value with the last 2 elements in `y` in reverse order
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the dot product of every other element in `x` starting from the second element with the last 2 elements in `y` in reverse order
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,216 @@
+var Complex128Array = require( '@stdlib/array/complex128' );
+
+var x = new Complex128Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 ] );
+var y = new Complex128Array( [ 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0 ] ); // eslint-disable-line max-len
```
```suggestion
var y = new Complex128Array( [ 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,216 @@
+-   **offsetY**: starting index for `y`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the dot product of every other value in `x` starting from the second value with the last 2 elements in `y` in reverse order
+
```
```suggestion

<!-- eslint-disable max-len -->

```

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-blas-base-zdotu

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/benchmark/benchmark.js:
```diff
@@ -0,0 +1,107 @@
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			d = zdotu( x.length, x, 1, y, 1 );
+			if ( isnan( d ) ) {
```
This isn't a valid check. You need to check, e.g., the real component.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/benchmark/benchmark.js:
```diff
@@ -0,0 +1,107 @@
+			}
+		}
+		b.toc();
+		if ( isnan( d ) ) {
```
Here, you can check the imaginary component (e.g., `isnan( imag( d ) )`).

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,107 @@
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			d = zdotu( x.length, x, 1, 0, y, 1, 0 );
+			if ( isnan( d ) ) {
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/repl.txt:
```diff
@@ -0,0 +1,109 @@
+    Returns
+    -------
+    out: Complex128
+        Scalar constant.
```
```suggestion
        Dot product.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/repl.txt:
```diff
@@ -0,0 +1,109 @@
+    vectors using alternative indexing semantics.
+
+    While typed array views mandate a view offset based on the underlying
+    buffer, the offset parameters support indexing based on a starting index.
```
```suggestion
    buffer, offset parameters support indexing based on starting indices.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/repl.txt:
```diff
@@ -0,0 +1,109 @@
+    Returns
+    -------
+    out: Complex128
+        Second input array.
```
```suggestion
        Dot product.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+*/
+interface Routine {
+	/**
+	* Calculates the dot product of vectors `x` and `y`.
```
Would you mind updating all the descriptions to use consistent language? s/Calculates/Computes/ here and everywhere else.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/test.ts:
```diff
@@ -0,0 +1,249 @@
+
+// TESTS //
+
+// The function returns Complex128...
```
```suggestion
// The function returns a complex number...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/test.ts:
```diff
@@ -0,0 +1,249 @@
+	zdotu( x.length, x, 1, y, 1, 10 ); // $ExpectError
+}
+
+// Attached to main export is an `ndarray` method which returns a Complex128...
```
```suggestion
// Attached to main export is an `ndarray` method which returns a complex number...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/lib/ndarray.js:
```diff
@@ -0,0 +1,73 @@
+	ix = offsetX;
+	iy = offsetY;
+	for ( i = 0; i < N; i++ ) {
+		dot = muladd( x.get( ix ), y.get( iy ), dot );
```
This isn't correct. We should not be using the accessors. We should reinterpret the complex input arrays as `Float64Arrays` of interleaved real and imaginary components. And then we should use `muladd.strided`. For reference, see https://github.com/stdlib-js/stdlib/blob/02838049b98a8ae3c595714e397364aace06f7b9/lib/node_modules/%40stdlib/blas/base/zaxpy/lib/ndarray.js#L79.

kgryte replied:
Instead of `muladd.assign`, you want to use `muladd.strided`: https://github.com/stdlib-js/stdlib/tree/02838049b98a8ae3c595714e397364aace06f7b9/lib/node_modules/%40stdlib/complex/float64/base/mul-add#muladdstrided-alpha-sa-oa-x-sx-ox-y-sy-oy-out-so-oo-

This allows you to have the output real and imaginary components stored in a `Float64Array`. To return a complex number, you then do `new Complex128( out[ 0 ], out[ 1 ] )`.

kgryte replied:
The reason we shouldn't be using accessors and constantly materializing `Complex128` instances is that all of this is happening in a hot loop. Instead, prefer to operate on the individual components directly.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.ndarray.js:
```diff
@@ -0,0 +1,239 @@
+	y = new Complex128Array( [ 2.0, 6.0, -1.0, -4.0, 8.0, 8.0, 2.0, -3.0 ] );
+
+	dot = zdotu( x.length, x, 1, 0, y, 1, 0 );
+	t.strictEqual(isSameComplex128( dot, new Complex128( 3, 70 ) ), true, 'returns expected value' );
```
```suggestion
	t.strictEqual(isSameComplex128( dot, new Complex128( 3.0, 70.0 ) ), true, 'returns expected value' );
```

To indicate that the input values are double, we should add decimals here and everywhere else in these test files when creating `Complex128` instances.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.ndarray.js:
```diff
@@ -0,0 +1,239 @@
+		7.0,  // 0
+		8.0,  // 0
+		9.0,  // 1
+		10.0, // 1
```
These annotations are not correct. The `0` and `1` comments should be reversed, as the strided for `y` is `-1` at L236.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.zdotu.js:
```diff
@@ -0,0 +1,217 @@
+	y = new Complex128Array( [ 2.0, 6.0, -1.0, -4.0, 8.0, 8.0, 2.0, -3.0 ] );
+
+	dot = zdotu( 4, x, 1, y, 1 );
+	t.strictEqual( isSameComplex128( dot, new Complex128( 3, 70 ) ), true, 'returns expected value' );
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.zdotu.js:
```diff
@@ -0,0 +1,217 @@
+	t.end();
+});
+
+tape( 'if provided an `N` parameter less than or equal to `0`, the function returns `0 + 0i`', function test( t ) {
```
```suggestion
tape( 'if provided an `N` parameter less than or equal to `0`, the function returns `0`', function test( t ) {
```

Consistency with other test file.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.zdotu.js:
```diff
@@ -0,0 +1,217 @@
+		6.0  // 1
+	]);
+	y = new Complex128Array([
+		7.0,  // 0
```
Same comment regarding the comments.

kgryte reviewed (changes requested):
This PR needs a bit of clean-up still.

@ShabiShett07 Also, in addition to the main export and the `ndarray` method, we should add an `assign` method with the following signature:

```javascript
zdotu.assign( N, x, strideX, offsetX, y, strideY, offsetY, out, offsetOut );
```

In this case, the real and imaginary components of the complex number output will be written to `out` which should be a `Complex128Array` at the index `offsetOut`.

The `ndarray` method can then wrap the `assign` API and you can use a preallocated `Complex128Array` workspace array which you then pass as input to `assign`. When returning from `ndarray`, you then do `workspace.get( 0 )`, which will materialize a new `Complex128` value.

Similarly, the main export can also wrap the `assign` API.

Why are we doing this? Because in `zgemm`, you will want to use the `assign` API of `zdotu`, not the `ndarray` API. The assign API will allow you to avoid materializing complex number instances in hot loops.

Commit:
chore: resolve suggestions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Commit:
chore: update documentation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
refactor: add assign API

kgryte commented:
@ShabiShett07 Would you mind resolving the lint failures in this PR? Cheers!

Commit:
fix: lint errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
refactor: use assign API

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

MeKaustubh07 committed:
Merge remote-tracking branch 'upstream/develop' into pr/6613

MeKaustubh07 committed:
fixed description

MeKaustubh07 committed:
Use String format instead of Concatination

MeKaustubh07 commented:
/stdlib merge

MeKaustubh07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

MeKaustubh07 commented:
@kgryte ready for the review

gururaj1512 commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,218 @@
+
+-   **N**: number of indexed elements.
+-   **x**: input [`Complex128Array`][@stdlib/array/complex128].
+-   **strideX**: stride kength for `x`.
```
```suggestion
-   **strideX**: stride length for `x`.
```

gururaj1512 commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,218 @@
+-   **x**: input [`Complex128Array`][@stdlib/array/complex128].
+-   **strideX**: stride kength for `x`.
+-   **y**: input [`Complex128Array`][@stdlib/array/complex128].
+-   **strideY**: stride kength for `y`.
```
```suggestion
-   **strideY**: stride length for `y`.
```

MeKaustubh07 committed:
Applied Suggestions

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/benchmark/benchmark.js:
```diff
@@ -0,0 +1,110 @@
+			}
+		}
+		b.toc();
+		if ( isnan( imag( d ) ) ) {
```
```suggestion
		if ( isnan( imag( z ) ) ) {
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,110 @@
+			}
+		}
+		b.toc();
+		if ( isnan( imag( d ) ) ) {
```
```suggestion
		if ( isnan( imag( z ) ) ) {
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+	( N: number, x: Complex128Array, strideX: number, y: Complex128Array, strideY: number ): Complex128;
+
+	/**
+	* Computes the dot product of vectors `x` and `y` using alternative indexing semantics.
```
```suggestion
	* Computes the dot product of strided arrays `x` and `y` using alternative indexing semantics.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+}
+
+/**
+* Computes the dot product of two double-precision complex floating-point vectors.
```
```suggestion
* Computes the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/repl.txt:
```diff
@@ -0,0 +1,109 @@
+
+{{alias}}( N, x, strideX, y, strideY )
+    Computes the dot product of two double-precision complex floating-point
+    vectors.
```
```suggestion
    strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/repl.txt:
```diff
@@ -0,0 +1,109 @@
+
+{{alias}}.ndarray( N, x, strideX, offsetX, y, strideY, offsetY )
+    Computes the dot product of two double-precision complex floating-point
+    vectors using alternative indexing semantics.
```
```suggestion
    strided arrays using alternative indexing semantics.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/lib/assign.js:
```diff
@@ -0,0 +1,90 @@
+// MAIN //
+
+/**
+* Computes the dot product of two double-precision complex floating-point vectors.
```
```suggestion
* Computes the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/lib/index.js:
```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+/**
+* BLAS level 1 routine to compute the dot product of two double-precision complex floating-point vectors.
```
```suggestion
* BLAS level 1 routine to compute the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/lib/ndarray.js:
```diff
@@ -0,0 +1,63 @@
+// MAIN //
+
+/**
+* Computes the dot product of two double-precision complex floating-point vectors.
```
```suggestion
* Computes the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/lib/zdotu.js:
```diff
@@ -0,0 +1,57 @@
+// MAIN //
+
+/**
+* Computes the dot product of two double-precision complex floating-point vectors.
```
```suggestion
* Computes the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.js:
```diff
@@ -0,0 +1,82 @@
+		'@stdlib/utils/try-require': tryRequire
+	});
+
+	t.strictEqual( zdotu, mock, 'returns expected value' );
```
```suggestion
	t.strictEqual( zdotu, mock, 'returns native implementation' );
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.js:
```diff
@@ -0,0 +1,82 @@
+		'@stdlib/utils/try-require': tryRequire
+	});
+
+	t.strictEqual( zdotu, main, 'returns expected value' );
```
```suggestion
	t.strictEqual( zdotu, main, 'returns JavaScript implementation' );
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/docs/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+*/
+interface Routine {
+	/**
+	* Computes the dot product of vectors `x` and `y`.
```
```suggestion
	* Computes the dot product of strided arrays `x` and `y`.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.ndarray.js:
```diff
@@ -0,0 +1,239 @@
+	t.end();
+});
+
+tape( 'the function calculates the dot product of two double-precision complex floating-point vectors `x` and `y`', function test( t ) {
```
```suggestion
tape( 'the function calculates the dot product of strided arrays `x` and `y`', function test( t ) {
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/test/test.zdotu.js:
```diff
@@ -0,0 +1,217 @@
+	t.end();
+});
+
+tape( 'the function calculates the dot product of two double-precision complex floating-point vectors `x` and `y`', function test( t ) {
```
```suggestion
tape( 'the function calculates the dot product of strided arrays `x` and `y`', function test( t ) {
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/package.json:
```diff
@@ -0,0 +1,82 @@
+{
+  "name": "@stdlib/blas/base/zdotu",
+  "version": "0.0.0",
+  "description": "Compute the dot product of two double-precision complex floating-point vectors.",
```
```suggestion
  "description": "Compute the dot product of two double-precision complex floating-point strided arrays.",
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,218 @@
+
+# zdotu
+
+> Calculate the dot product of two double-precision complex floating-point vectors.
```
```suggestion
> Calculate the dot product of two double-precision complex floating-point strided arrays.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,218 @@
+
+#### zdotu( N, x, strideX, y, strideY )
+
+Computes the dot product of vectors `x` and `y`.
```
```suggestion
Computes the dot product of strided arrays `x` and `y`.
```

MeKaustubh07 commented on lib/node_modules/@stdlib/blas/base/zdotu/README.md:
```diff
@@ -0,0 +1,218 @@
+
+#### zdotu.ndarray( N, x, strideX, offsetX, y, strideY, offsetY )
+
+Computes the dot product of `x` and `y` using alternative indexing semantics.
```
```suggestion
Computes the dot product of strided arrays `x` and `y` using alternative indexing semantics.
```

MeKaustubh07 committed:
Apply suggestions from code review

Co-authored-by: Kaustubh Patange <kaustubh.mp007@gmail.com>
Signed-off-by: Kaustubh Patange <kaustubh.mp007@gmail.com>
